lib/ansible/modules/cloud/amazon/aws_acm.py

```diff
@@ -52,6 +52,7 @@
     this module will delete all ACM resources in this AWS region with a corresponding C(Name) tag.
     If there are none, it will succeed without effect.
   - Note that this may not work properly with keys of size 4096 bits, due to a limitation of the ACM API.
+  - This module uses the C(cryptography) python library to interact with OpenSSL. (C(pip install cryptography))
```

Comment:
You need to add it to the module requirements, then.

Reply:
You also need to make sure that `cryptography` is actually there, and fail gently if it is not. Right now, the module will crash if `cryptography` is not installed when `get_fingerprint_from_pem_cert` is called.

Reply:
And finally, the most important point, `get_fingerprint_from_pem_cert` won't help you for certificate chains at all! (Except if they have length 1.)

Reply:
1. I just added it to the requirements

2. Ah true. I thought the module_util handles that. But upon closer inspection it just carries on. Can you show me an example of a graceful failure? Should I try to `import cryptography` inside the main handler with a `try`?

3. Ugh, yes that's true. I can't see any function in `cryptography` or `crypto.py` which handles chains. Damn. I don't actually know much about the syntax. Last I saw, there's some extra stuff at the start of each cert. Hmm, maybe I really do need to look up the X.509 standard and write my own proper parser.

Reply:
For 2: see https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_privatekey.py#L230-L245 and https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/openssl_privatekey.py#L683-L685

For 3: that's because there is no such function :) Chains are simply concatenated PEM files. Definitely don't write your own parser, please :) That's never a good idea, especially not for complex beasts such as X.509. (Guess how many security vulnerabilities have been introduced in software when people wrote their own X.509 parser in C/C++ or other memory-unsafe languages...)

The best approach is to look for lines starting with `-----`, and just concatenate the lines between them (between the first and second, between the third and fourth, ...) and concatenate them (with spacing and newlines removed) - so similarly as you were doing in an earlier version of this PR. Or you use code similarly to https://github.com/ansible/ansible/blob/devel/lib/ansible/modules/crypto/acme/acme_certificate.py#L709-L723 to split up the chain into PEM certificates (and then process them with `cryptography` or `ssl.PEM_cert_to_DER_cert` (https://docs.python.org/3/library/ssl.html#ssl.PEM_cert_to_DER_cert, no extra dependency needed).

Reply:
Ok, one of the real certificate chains I'm dealing with looks like

```
subject=/C=AU/ST=Victoria/L=Melbourne/O=Telstra Corporation Limited/OU=Telstra Energy/CN=dev.energy.inside.telstra.com
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Global SSL ICA G3
-----BEGIN CERTIFICATE-----
MIIIHTCCB...g6Z5Q=
-----END CERTIFICATE-----

subject=/C=BM/O=QuoVadis Limited/CN=QuoVadis Global SSL ICA G3
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
-----BEGIN CERTIFICATE-----
MIIGF...KmnySglYc
-----END CERTIFICATE-----

subject=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
issuer=/C=BM/O=QuoVadis Limited/CN=QuoVadis Root CA 2 G3
-----BEGIN CERTIFICATE-----
MIIFYDCCA...mD3n+M
-----END CERTIFICATE-----

```

Are you saying I should ignore the `subject` and `issues` lines?

Reply:
Can I just compare two DER strings, after remove whitespace?

Reply:
I've rewritten my code, using that `ssl` library function to convert each cert to DER. I'm assuming I can just compare DER with `==`.

Please have a look and let me know if my solution is good.